docs/design-docs/02-custom-root-reconciler-clusterrole.md

```diff
@@ -18,65 +24,130 @@ This proposal introduces a mechanism to customize this behavior.
 on `spec.overrides` for a `RootSync`:
 
 ```yaml
+kind: RootSync
+metadata:
+  name: my-root-sync
+  namespace: config-management-system
 spec:
   overrides:
-    clusterRole:
-      name: my-custom-role
+    roleRefs:
+    - kind: ClusterRole # Creates a ClusterRoleBinding bound to ClusterRole my-cluster-role
+      name: my-cluster-role
+    - kind: ClusterRole # Creates a RoleBinding in my-tenant-namespace bound to ClusterRole tenant-cluster-role
+      name: tenant-cluster-role
+      namespace: my-tenant-namespace
+    - kind: Role # Creates a RoleBinding in my-tenant-namespace bound to Role my-tenant-role
+      name: my-tenant-role
+      namespace: my-tenant-namespace
 ```
 
-This role is defaulted to `cluster-admin` in order to stay backwards compatible.
+For `RootSync` objects, the default behavior will create a ClusterRoleBinding
+to `cluster-admin` in order to stay backwards compatible.
+
+For convenience, `RootSync` reconcilers will also be bound to a base `ClusterRole`
+which gives the reconciler the permissions for basic functionality. This will
+be comparable to the pre-existing base ClusterRole for `RepoSync` reconcilers,
+which includes permissions such as status writing on `RepoSync` objects. Leaving
+this permission to the user to manage would create unneeded toil.
+
+### `RepoSync`s
 
-For the case where a single `ClusterRole` is not expressive enough to configure the
-permissions a user want, you can instead set `disabled: true` on the override object:
+For `RepoSync` objects, the `namespace` field will not be present on `roleRefs`.
+Instead, the `namespace` will always be derived from the `RepoSync` object itself.
+This follows the assumption that `RepoSync` objects are always scoped to a single
+Namespace.
 
 ```yaml
+kind: RepoSync
+metadata:
+  name: my-repo-sync
+  namespace: example-ns
 spec:
   overrides:
-    clusterRole:
-      disabled: true
+    roleRefs:
+    - kind: ClusterRole # Creates a RoleBinding in example-ns bound to ClusterRole my-cluster-role
+      name: my-cluster-role
+    - kind: Role # Creates a RoleBinding in example-ns bound to Role my-tenant-role
+      name: my-tenant-role
 ```
-This disables creating the `ClusterRoleBinding` entirely.
 
-If both `disabled: true` and a custom name is specified, `disabled` "wins" and no binding
-is created.
+### Lifecycle management
+
+Given this API provides configuration for a list of RoleRefs, it requires some form
+of lifecycle management to clean up stale bindings. For example if a user removes
+one roleRef from the list of roleRefs, they would reasonably expect that the
+binding will be garbage collected.
+
+This essentially requires for the `reconciler-manager` to be able to track an
+inventory of bindings that were previously created for a given `RootSync` or `RepoSync`.
+This can be accomplished by applying a label whenever a new binding is created,
+and then querying using a label selector on subsequent reconciliation loops.
```

Comment:
Can we end up in a situation where the `reconciler-manager` will not be able to delete the bindings (maybe because it no longer has permissions)? If so, how is the issue surfaced to users?

Reply:
I don't think that should ever happen, unless a user has a standalone Config Sync installation (no managing operator) and manually modifies the reconciler-manager ClusterRole/ClusterRoleBinding.

If such a error did happen, I expect the behavior would be the same regardless of which managed resource it's erroring on (e.g. ServiceAccount, Deployment, RoleBinding, etc). I haven't tested this to verify, but my expectation is that the reconciler-manager would stall and surface the error on the RSync.

Reply:
In general, when the reconciler-manager or reconciler cannot delete things that it manages, it errors and that error is surfaced in the RSync status.